crates/sui-types/src/sui_system_state/mod.rs

```diff
@@ -68,6 +72,94 @@ impl SuiSystemStateWrapper {
             type_params: vec![],
         }
     }
+
+    pub fn advance_epoch_safe_mode<S>(
+        &self,
+        params: &AdvanceEpochParams,
+        object_store: &S,
+        protocol_config: &ProtocolConfig,
+    ) -> Object
+    where
+        S: ObjectStore,
+    {
+        let id = self.id.id.bytes;
+        let mut field_object = get_dynamic_field_object_from_store(object_store, id, &self.version)
+            .expect("Dynamic field object of wrapper should always be present in the object store");
+        let move_object = field_object
+            .data
+            .try_as_move_mut()
+            .expect("Dynamic field object must be a Move object");
+        match self.version {
+            1 => {
+                Self::advance_epoch_safe_mode_impl::<SuiSystemStateInnerV1>(
+                    move_object,
+                    params,
+                    protocol_config,
+                );
+            }
+            2 => {
+                Self::advance_epoch_safe_mode_impl::<SuiSystemStateInnerV1>(
+                    move_object,
+                    params,
+                    protocol_config,
+                );
+            }
```

Comment:
We merged a `SuiSystemStateInnerV2` with the extra `min_validator_count` system parameter a few days ago. That will go in as framework upgrade in protocol version 3 right? So I believe we should have a pattern matching `3` here that calls `Self::advance_epoch_safe_mode_impl::<SuiSystemStateInnerV2>`. I'm surprised that no tests caught this....Or did I miss something?

Reply:
Ouch. 2 should map to SuiSystemStateInnerV2.
I wonder if there is a better way to code these mappings so that it's less error prone

Reply:
Good catch!

Reply:
Hmm why should 2 map to `SuiSystemStateInnerV2`? We are already at version 2 in testnet now and when we upgrade the framework with the new code we'll be at version 3?

Reply:
This is the sui system state type version, which can be different from the protocol version

Reply:
Got it. 
> I wonder if there is a better way to code these mappings so that it's less error prone

Deserialization in `advance_epoch_safe_mode_impl` should fail if we provide the wrong type right?

Reply:
Yeah but then to test it we would need to enter safe mode in v2.
The current test enters safe mode in v1 I think. So maybe the only way is to stress it more in tests

Reply:
Fixed.

Reply:
> The current test enters safe mode in v1 I think.

If we are thinking about the same (only) safe mode test, that one uses mock sui system package and uses `SimTestSuiSystemStateInnerV1`.

Reply:
Ah so there is actually no good way to test this today

Reply:
Unless I manually introduce a bug

Reply:
Maybe for that safe mode test, we should indeed do what @amnn originally suggested under that PR, which is using the current sui framework as base, and programmatically change one line to introduce an abort.

Reply:
This way we get to test safe mode for the current system. What do you think? I'll look into it

Reply:
It may also be possible to inject a failure/error right after the Move call in Rust?